packages/flutter_tools/test/general.shard/runner/runner_test.dart

```diff
@@ -0,0 +1,152 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:file/memory.dart';
+import 'package:flutter_tools/runner.dart' as runner;
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/io.dart' as io;
+import 'package:flutter_tools/src/base/common.dart';
+import 'package:flutter_tools/src/cache.dart';
+import 'package:flutter_tools/src/reporting/reporting.dart';
+import 'package:flutter_tools/src/runner/flutter_command.dart';
+import 'package:platform/platform.dart';
+
+import '../../src/common.dart';
+import '../../src/context.dart';
+
+
+void main() {
+  group('runner', () {
+    setUp(() {
+      runner.crashFileSystem = MemoryFileSystem();
+      io.setExitFunctionForTests((int _) { throw 'test exit';});
+      Cache.disableLocking();
+    });
+
+    tearDown(() {
+      runner.crashFileSystem = const LocalFileSystem();
+      io.restoreExitFunction();
+      Cache.enableLocking();
+    });
+
+    testUsingContext('error handling', () async {
+      final Completer<void> completer = Completer<void>();
+      // runner.run() asynchronously calls the exit function set above, so we
+      // catch it in a zone.
+      unawaited(runZoned<Future<void>>(() {
+          unawaited(runner.run(
+            <String>['test'],
+            <FlutterCommand>[
+              CrashingFlutterCommand(),
+            ],
+            flutterVersion: '[user-branch]/',
+            reportCrashes: true,
+          ));
+          return null;
+        },
+        onError: (Object error) {
+          expect(error, 'test exit');
+          completer.complete();
+        }));
+      await completer.future;
+      final CrashingUsage crashingUsage = flutterUsage;
+
+      // This is the main check of this test.
+      //
+      // We are checking that, even though crash reporting failed with an
+      // exception on the first attempt, the second attempt tries to report the
+      // *original* crash, and not the crash from the first crash report
+      // attempt.
+      expect(crashingUsage.sentException, 'runCommand');
```

Comment:
awesome test! Do you know what scenarios can cause this issue?

Reply:
Yes. In `runner.run()`, if the call to `_handleToolError()` in the catch block throws an exception, it will be caught by the `onError` callback to `runZoned()`.

Reply:
I'm looking at the implementation of `_handleToolError()`. It looks like some state changed between the two invocations to `_handleToolError()`?